### ISSUE TITLE
Task.Delay() waits shorter than expected in .NET 7 and 8

### ISSUE THREAD

**Issue by MaciekSwiszczowski** [dotnet/runtime#87112]
### Description

I've found that Task.Delay(1) and Task.Delay(2) may take shorter than the expected 1 or 2 milliseconds. This can be observed in the reproduction app. These short delays can be also measured in Visual Studio, when running the application with Debug->Performance Profiler->.NET async.

The issue was reproduced on a few Windows 11 machines. It doesn't appear on lower .NET version.

### Reproduction Steps

``` c#
await DelayTest(1, 1000);
await DelayTest(2, 10000);

    private static async Task DelayTest(int milliseconds, int iterationsCount)
    {
        var stopWatch = new Stopwatch();
        stopWatch.Start();
        var deltas = new TimeSpan[iterationsCount];
        var tick = stopWatch.Elapsed;

        for (var j = 0; j < iterationsCount; j++)
        {
            await Task.Delay(milliseconds);

            deltas[j] = stopWatch.Elapsed - tick;
            tick = stopWatch.Elapsed;

            Thread.SpinWait(10000 + 5 * j);
        }

        var shortSpans =
            deltas
                .Where(i => i.TotalMilliseconds <= milliseconds)
                .OrderBy(i => i)
                .Select(i => i.TotalMilliseconds)
                .ToList();

        Console.WriteLine();
        Console.WriteLine(shortSpans.Any() ? string.Join(" ", shortSpans) : "No issues found");
    }

### Expected behavior

`await Task.Delay(expected_milliseconds)` should wait at least `expected_milliseconds`.

### Actual behavior

`await Task.Delay(expected_milliseconds)` sometimes waits shorter than  `expected_milliseconds`

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

tested on:
- Windows 11,
-  .NET 7 & 8, 
- x64 / AnyCPU

### Other information

_No response_

**Comment by stephentoub** [comment]
I've not tried to repro, but Task.Delay is just wrapping Timer, so presumably you see the same behavior with it directly.

**Comment by HypsyNZ** [comment]
I'm glad that the appcontext switch to use this "new" timer doesn't even work then.

**Comment by MaciekSwiszczowski** [comment]
Sure, but that should be documented at least.

**Comment by HypsyNZ** [comment]
Its not a timer, its a delay to break up busy work.

**Comment by HypsyNZ** [comment]
Its not a timer, its a delay to break up busy work.

If you need a timer use a timer.

```cs
while(true)
{
    // work
    Task.Delay(1);
}
```

If you are doing this you should be using a timer call back with proper exception handling anyway.

**Comment by RedworkDE** [comment]
1. Yes this also happens when using Threading.Timer and Timers.Timer, not with Thread.Sleep tho
2. This can happen for any duration not just 1/2 ms
3. For some durations it is pretty likely that the timer is early, eg for 15ms (and small multiples) 10-20% of the waits have ended early
4. The timer is never more than 1ms early
5. What happens is that the timer internally works at ms resolution only and the timer effectively starts at the beginning of the current millisecond and not the time of the call, thus when staring a timer for `t`ms at the end of millisecond `n` and the internal timer fires at the start of millisecond `n+t` the timer will fire after having waited only `t-1`ms
6. A simplified test project (It will only show the issue for some durations, but is way faster):
```cs
async Task DelayTest(int milliseconds, int iterationsCount)
{
    var stopWatch = new System.Diagnostics.Stopwatch();
    var n = 0;

    for (var j = 0; j < iterationsCount; j++)
    {
        stopWatch.Restart();
        await Task.Delay(milliseconds);
        stopWatch.Stop();

        if (stopWatch.Elapsed.TotalMilliseconds < milliseconds)
            n++;
    }

    Console.WriteLine($"{milliseconds}ms: {n}/{iterationsCount} {(float)n / iterationsCount:P}");
}

for (int i = 1; i < 35; i++)
{
    await DelayTest(i, 1_00);
}
```

**Comment by MaciekSwiszczowski** [comment]
Thanks for the explanation!

**Comment by danmoseley** [comment]
@kouvel  is there any work here or just docs?

**Comment by kouvel** [comment]
Currently, none of the system clocks used (in the timer thread for waits, or in timer due time tracking) have sub-ms precision. Although the due time tracking could be updated to use a high-precision clock, it is a lot more expensive, and there can be times when the timer thread wakes up only to find that none of the timers' due times have expired and go back to waiting again. The current behavior seems reasonable for most cases. I think it would be good to update the docs to clarify.

**Comment by Clockwork-Muse** [comment]
@MaciekSwiszczowski - Are you expecting the delay to be **at least** the indicated 1 or 2 ms?

(Historically, most things that were delayed, or were dealing with lower-resolution timers, or similar functions, had a 60hz resolution - including things like getting the time from the clock!)

Otherwise, this feels like a more specific form of "your application must account for timing issues arising from the vagaries of CPU scheduling and timer resolution".  Which is a well known (or should be) limitation...

**Comment by MaciekSwiszczowski** [comment]
@Clockwork-Muse Yes, my understanding of Task.Delay() was that it should wait at least the expected amount of milliseconds. But if I needed such a resolution in production code, I'd make research.

I've just observed this behavior of waiting less than expected, and I've checked that I can't reproduce it for .NET versions 5 and 6, so I thought it's a regression issue.

**Comment by kouvel** [comment]
Prior to .NET 7 a different Windows API was used for timer due time tracking, which may use a slightly different clock, though it also apparently has similar precision and would have similar issues. I don't see any reason why the prior implementations would have stronger guarantees.

**Comment by RedworkDE** [comment]
IMO this is just an off-by-one error and replacing this `<=` with a `<` should (I didn't test it) fix the issue by ensuring the minimum wait is until the end of the target millisecond: https://github.com/dotnet/runtime/blob/10222f94e5d89b19959117dd9e5b8576ef63f878/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs#L211

**Comment by kouvel** [comment]
The resolution of the timers involved are by default ~16 ms, so a difference of 1 ms would not be sufficient to make any guarantees

**Comment by kouvel** [comment]
Even if the timers were to be configured to have a 1 ms resolution, the timer tick could occur perhaps ideally at most 1 ms earlier than requested.

**Comment by kouvel** [comment]
> replacing this `<=` with a `<`

That would probably work if all clocks are configured to have a 1 ms resolution to solve this issue, but otherwise in most other cases it may introduce cases where the timer thread wakes up and doesn't find any timers to fire due to the same change.

**Comment by Heshyo** [comment]
Maybe linked to https://github.com/dotnet/runtime/issues/45585 ?
